checks/shell_download_validate.go

```diff
@@ -0,0 +1,628 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"mvdan.cc/sh/v3/syntax"
+)
+
+// ErrParsingDockerfile indicates a problem parsing the dockerfile.
+var ErrParsingDockerfile = errors.New("file cannot be parsed")
+
+// ErrParsingShellCommand indicates a problem parsing a shell command.
+var ErrParsingShellCommand = errors.New("shell command cannot be parsed")
+
+// List of interpreters.
+var interpreters = []string{
+	"sh", "bash", "dash", "ksh", "python",
+	"perl", "ruby", "php", "node", "nodejs", "java",
+	"exec",
+}
+
+func isBinaryName(expected, name string) bool {
+	return strings.EqualFold(path.Base(name), expected)
+}
+
+// see https://serverfault.com/questions/226386/wget-a-script-and-run-it/890417.
+func isDownloadUtility(cmd []string) bool {
+	if len(cmd) == 0 {
+		return false
+	}
+	// Note: we won't be catching those if developers have re-named
+	// the utility.
+	// Note: wget -O - <website>, but we don't check for that explicitly.
+	utils := [3]string{"curl", "wget", "gsutil"}
+	for _, b := range utils {
+		if isBinaryName(b, cmd[0]) {
+			return true
+		}
+	}
+
+	// aws s3api get-object.
+	// https://docs.aws.amazon.com/AmazonS3/latest/userguide/download-objects.html.
+	if isBinaryName("aws", cmd[0]) {
+		if len(cmd) >= 3 && strings.EqualFold("s3api", cmd[1]) && strings.EqualFold("get-object", cmd[2]) {
+			return true
+		}
+	}
+	return false
+}
+
+func getWgetOututFile(cmd []string) (pathfn string, ok bool, err error) {
+	if isBinaryName("wget", cmd[0]) {
+		for i := 1; i < len(cmd)-1; i++ {
+			// Find -O output, or use the basename from url.
+			if strings.EqualFold(cmd[i], "-O") {
+				return cmd[i+1], true, nil
+			}
+		}
+
+		// Could not find -O option, use the url's name instead.
+		for i := 1; i < len(cmd); i++ {
+			if !strings.HasPrefix(cmd[i], "http") {
+				continue
+			}
+
+			u, err := url.Parse(cmd[i])
+			if err != nil {
+				return "", false, fmt.Errorf("url.Parse: %w", err)
+			}
+			return path.Base(u.Path), true, nil
+		}
+	}
+	return "", false, nil
+}
+
+func getGsutilOututFile(cmd []string) (pathfn string, ok bool, err error) {
+	if isBinaryName("gsutil", cmd[0]) {
+		for i := 1; i < len(cmd)-1; i++ {
+			if !strings.HasPrefix(cmd[i], "gs://") {
+				continue
+			}
+			pathfn := cmd[i+1]
+			if filepath.Clean(filepath.Dir(pathfn)) == filepath.Clean(pathfn) {
+				// Directory.
+				u, err := url.Parse(cmd[i])
+				if err != nil {
+					return "", false, fmt.Errorf("url.Parse: %w", err)
+				}
+				return filepath.Join(filepath.Dir(pathfn), path.Base(u.Path)), true, nil
+			}
+
+			// File provided.
+			return pathfn, true, nil
+		}
+	}
+	return "", false, nil
+}
+
+func getAwsOututFile(cmd []string) (pathfn string, ok bool, err error) {
```

Comment:
Aws should be AWS per https://github.com/golang/go/wiki/CodeReviewComments#initialisms